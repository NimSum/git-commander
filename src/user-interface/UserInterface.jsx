import React, { Component } from 'react'
import { ChallengeCard } from './ChallengeCard';
import { ChargeBar } from './ChargeBar';

export class UserInterface extends Component {
  constructor(props) {
    super(props)
    this.state = {
      currentQuestion: {},
      currDifficulty: 1,
      userAnswer: ''
    }
  }

  generateChallenge() {
    const randomQuestion = this.props.challenges
    .filter(challenge => challenge.difficulty === this.state.currDifficulty.toString())
    .pop();
    this.setState( {currentQuestion: randomQuestion} )
  }

  componentDidMount() {
    this.generateChallenge()
  }

  verifyAnswer = e => {
    e.preventDefault();
    if (this.state.currentQuestion.answer === this.state.userAnswer) {
      e.target.reset();
      this.changeDifficulty();
      this.props.nextRound();
      this.generateChallenge();
    }
  }

  changeDifficulty() {
    if (this.state.currDifficulty < 5) {
      this.setState({
        currDifficulty: this.state.currDifficulty + Math.round(Math.random() + .05)
      })
    }
  }

  handleChange = e => {
    this.setState({ userAnswer: e.target.value })
  }

  render() {
    return (
      <aside className="user-interface">
        <h2>Commander { this.props.playerName }</h2>
        < ChallengeCard challenge={this.state.currentQuestion}/>
        < ChargeBar />
        <form onSubmit={ this.verifyAnswer }>
          <label htmlFor="command-input">Command Center</label>
          <input onChange={ this.handleChange } id="command-input" type="text"/>
        </form>
      </aside>
    )
  }
}

export default UserInterface
