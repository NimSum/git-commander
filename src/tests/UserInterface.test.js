import React from 'react';
import UserInterface from '../user-interface/UserInterface';
import Adapter from 'enzyme-adapter-react-16';
import { shallow , configure } from 'enzyme';

configure({adapter: new Adapter()});

describe('User Interface' , () => {
  const mockData = [{
      "statement" : "Our Git spaceship needs to switch to a different branch to steer the ship!",
      "detail" : "Switch to the branch called \"steer-ship\" to avoid the obstacle",
      "answer" : "git checkout steer-ship", 
      "difficulty" : "1",
      "moreInfo" : "/////A link to where they could read more about this command",
      "id": "1"
    },{
      "statement" : "Our Git spaceship needs to switch to a different branch to steer the ship!",
      "detail" : "Switch to the branch called \"steer-ship\" to avoid the obstacle",
      "answer" : "git checkout steer-ship", 
      "difficulty" : "1",
      "moreInfo" : "/////A link to where they could read more about this command",
      "id": "1"
    }]
  
  const mockEvent = {
    target: { 
      value: 'git checkout steer-ship',
      reset: () => {}
    },
    preventDefault: () => {}
  };
  const mockNextRound = jest.fn();
  const mockProps = {    
    challenges: mockData,
    currentRound: 1
  };
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      < UserInterface 
        nextRound={mockNextRound}
        {...mockProps}
      />
    );
  });

  it('Should match render snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('Should automatically update states when mounted using generateChallenge', () => {
    expect(wrapper.state()).toEqual({
      currentQuestion: mockData[0],
      currDifficulty: 1,
      questionsByDiff: [mockData[0]],
      userAnswer: ''
    });
  })

  it('Should handle user command input', () => {
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('userAnswer')).toEqual('git checkout steer-ship');
  })

  it('Should verify user answer', () => {
    const spy = jest.spyOn(UserInterface.prototype, 'changeDifficulty');

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('userAnswer')).toEqual('git checkout steer-ship');

    wrapper.instance().verifyAnswer(mockEvent);
    expect(spy).toHaveBeenCalled();
    expect(mockNextRound).toHaveBeenCalled();
  })

  it('Should change difficulty', () => {
    const spy = jest.spyOn(wrapper.instance(), 'generateChallenge');
    
    wrapper.instance().changeDifficulty();
    expect(spy).toHaveBeenCalled();
  })


})