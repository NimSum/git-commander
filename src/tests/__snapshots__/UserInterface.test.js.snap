// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Interface Should match render snapshot 1`] = `
<aside
  className="user-interface"
>
  <h2>
    Commander 
  </h2>
  <ChallengeCard
    challenge={
      Object {
        "answer": "git checkout steer-ship",
        "detail": "Switch to the branch called \\"steer-ship\\" to avoid the obstacle",
        "difficulty": "1",
        "id": "1",
        "moreInfo": "/////A link to where they could read more about this command",
        "statement": "Our Git spaceship needs to switch to a different branch to steer the ship!",
      }
    }
  />
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="command-input"
    >
      Command Center
    </label>
    <input
      autoFocus={true}
      id="command-input"
      onChange={[Function]}
      type="text"
    />
  </form>
  <ChargeBar />
</aside>
`;
